[Ingest Pipelines] Add structure tree components

ElenaStoeva opened the pull request:
Addresses https://github.com/elastic/kibana/issues/227405


## Summary

This PR adds the Pipeline Structure Tree components and the relevant functionality in a new `ingest-pipelines` package. To test the component in Storybook, run `yarn storybook ingest_pipelines`. See design mocks in https://github.com/elastic/kibana/issues/227405.


<img width="613" height="740" alt="Screenshot 2025-07-11 at 14 54 11" src="https://github.com/user-attachments/assets/cbe1c3f3-2150-4548-a8bc-5705fa5cbf51" />


Note: There is no out-of-the-box EUI component that exactly matches the component from the design mocks so we are using EuiTreeView with added CSS for the tree nodes. However, this component doesn't have lines connecting the tree nodes and fo now we are creating the component without the lines. We could try adding them at the end. There is also an open issue (https://github.com/elastic/platform-ux-team/issues/780) for improving the EuiTreeView component.

ElenaStoeva committed:
[Ingest Pipelines] Add structure tree components

ElenaStoeva committed:
Add tree traversal, tree nodes, and tests

ElenaStoeva committed:
Add managed tooltip

ElenaStoeva committed:
Add styling

ElenaStoeva committed:
Small adjustments

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_packages --fix'

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/structure-tree

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/317665)
* Commit: 4eafaff9051a51776fdac3ed35d8c14a2761a87d

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/317665#0197f404-e7c9-4aec-b68d-d94db1158b89)

### History
* :broken_heart: [Build #317651](https://buildkite.com/elastic/kibana-pull-request/builds/317651) failed aa0ae050410081d931d40425c53732532391a597
* :broken_heart: [Build #317606](https://buildkite.com/elastic/kibana-pull-request/builds/317606) failed e01f8e50c80d7fd0c9e992f12fc4a84131cc9f33

cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317665","number":317665,"commit":"4eafaff9051a51776fdac3ed35d8c14a2761a87d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317651","number":317651,"commit":"aa0ae050410081d931d40425c53732532391a597"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317606","number":317606,"commit":"e01f8e50c80d7fd0c9e992f12fc4a84131cc9f33"}],"number":317665}
buildkite-pr-comment-->

ElenaStoeva committed:
Fix type errors

mattkime reviewed (changes requested):
The triangle should only show if there are children, seems to show every time. 

Overall looks good, just some minor requests.

sabarasaba reviewed (commented):
overall besides the styling discrepancies looks good, was wondering what was your plan for handling the load more button, it seems that as is we dont do anything about it but maybe you had some ideas already

ElenaStoeva committed:
Address feedback

ElenaStoeva commented:
Thank you all for the feedback! I made the following changes in the last commit:

- Displayed arrows only if the tree nodes have children
- The mock designs were slightly changed today (the Managed and Deprecated badges were changed) so I added the Deprecated icon to the tree
- Added the "More pipelines" node that @sabarasaba mentioned - will implement its onClick logic in a follow-up PR
- Refactored the code, added a component for the tree node labels
- Added logic for selecting a tree node, whose info will be displayed in the flyout to the right
- Improved the styling, simplified class names, added styling for selected tree node and for hovering

ElenaStoeva commented on x-pack/platform/packages/shared/ingest-pipelines/src/components/pipeline_structure_tree/styles.ts:
```diff
@@ -0,0 +1,25 @@
+    margin: 20px 0px;
+  }
+
+  .cssTreeNode-level-2,
```
Thanks! I simplified the class names in [ad3bc33](https://github.com/elastic/kibana/pull/226883/commits/ad3bc33ea1fa4b342c32fa10391e4538ff6f45b0).

ElenaStoeva commented on x-pack/platform/packages/shared/ingest-pipelines/.storybook/main.js:
```diff
@@ -0,0 +1,13 @@
+
+module.exports = {
+  ...defaultConfig,
+  stories: ['../**/*.stories.+(tsx|mdx)'],
```
Addressed in [ad3bc33](https://github.com/elastic/kibana/pull/226883/commits/ad3bc33ea1fa4b342c32fa10391e4538ff6f45b0)

ElenaStoeva commented on x-pack/platform/packages/shared/ingest-pipelines/src/components/pipeline_structure_tree/styles.ts:
```diff
@@ -0,0 +1,25 @@
+import { css } from '@emotion/react';
+
+export const styles = css`
+  [class*='cssTreeNode-level-'] {
```
Simplified the class names in [ad3bc33](https://github.com/elastic/kibana/pull/226883/commits/ad3bc33ea1fa4b342c32fa10391e4538ff6f45b0)

ElenaStoeva committed:
Fix typo and linter

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/yarn_deduplicate'

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318260)
* Commit: 7cc07a313fbc8284c5daf2cd334edd582dd96903

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/318260#0197f9e3-e351-4664-b48c-cb7629a4c51c)

### History
* :broken_heart: [Build #317817](https://buildkite.com/elastic/kibana-pull-request/builds/317817) failed d68908a3677c79a65705b05f6d6af3d2450b2e88
* :broken_heart: [Build #317665](https://buildkite.com/elastic/kibana-pull-request/builds/317665) failed 4eafaff9051a51776fdac3ed35d8c14a2761a87d
* :broken_heart: [Build #317651](https://buildkite.com/elastic/kibana-pull-request/builds/317651) failed aa0ae050410081d931d40425c53732532391a597
* :broken_heart: [Build #317606](https://buildkite.com/elastic/kibana-pull-request/builds/317606) failed e01f8e50c80d7fd0c9e992f12fc4a84131cc9f33

cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318260","number":318260,"commit":"7cc07a313fbc8284c5daf2cd334edd582dd96903"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317817","number":317817,"commit":"d68908a3677c79a65705b05f6d6af3d2450b2e88"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317665","number":317665,"commit":"4eafaff9051a51776fdac3ed35d8c14a2761a87d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317651","number":317651,"commit":"aa0ae050410081d931d40425c53732532391a597"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317606","number":317606,"commit":"e01f8e50c80d7fd0c9e992f12fc4a84131cc9f33"}],"number":318260}
buildkite-pr-comment-->

ElenaStoeva committed:
Fix type and lint errors

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318415)
* Commit: c0481649aa52ba73a9a7c04e92d60e70ad059cd2
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-226883)

### Failed CI Steps
* [Jest Tests #14](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fb9d-64d2-4c47-80b1-e7e421bf68bf)
* [Jest Tests #14](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fbbe-b22d-49dd-b123-f76a19f3e813)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fb9d-64d2-4c47-80b1-e7e421bf68bf) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fb9d-64d2-4c47-80b1-e7e421bf68bf/artifacts/0197fbbe-9754-4c06-8d5b-108d7d42c021) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates a node with a deprecated icon
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fbbe-b22d-49dd-b123-f76a19f3e813) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fbbe-b22d-49dd-b123-f76a19f3e813/artifacts/0197fbc2-f444-40c6-ad55-432eeec9b47e) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates a node with a deprecated icon
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fb9d-64d2-4c47-80b1-e7e421bf68bf) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fb9d-64d2-4c47-80b1-e7e421bf68bf/artifacts/0197fbbe-9754-461d-bb08-5f270a159e2e) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates a node with a managed icon
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fbbe-b22d-49dd-b123-f76a19f3e813) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fbbe-b22d-49dd-b123-f76a19f3e813/artifacts/0197fbc2-f445-4875-8371-d381f5d2c567) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates a node with a managed icon
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fb9d-64d2-4c47-80b1-e7e421bf68bf) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fb9d-64d2-4c47-80b1-e7e421bf68bf/artifacts/0197fbbe-9755-41fc-8215-230d2d758b4b) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates a single node with no children
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fbbe-b22d-49dd-b123-f76a19f3e813) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fbbe-b22d-49dd-b123-f76a19f3e813/artifacts/0197fbc2-f444-40b2-9a41-e09a1fa4f7db) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates a single node with no children
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fb9d-64d2-4c47-80b1-e7e421bf68bf) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fb9d-64d2-4c47-80b1-e7e421bf68bf/artifacts/0197fbbe-9755-4ffa-b5d0-e6241be71906) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates nested child nodes up to the max depth
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/318415#0197fbbe-b22d-49dd-b123-f76a19f3e813) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/318415/jobs/0197fbbe-b22d-49dd-b123-f76a19f3e813/artifacts/0197fbc2-f444-471b-bd35-13c9d6867065) Jest Tests #<span></span>14 / createTreeNodesFromPipelines creates nested child nodes up to the max depth

### History
* :broken_heart: [Build #318260](https://buildkite.com/elastic/kibana-pull-request/builds/318260) failed 7cc07a313fbc8284c5daf2cd334edd582dd96903
* :broken_heart: [Build #317817](https://buildkite.com/elastic/kibana-pull-request/builds/317817) failed d68908a3677c79a65705b05f6d6af3d2450b2e88
* :broken_heart: [Build #317665](https://buildkite.com/elastic/kibana-pull-request/builds/317665) failed 4eafaff9051a51776fdac3ed35d8c14a2761a87d
* :broken_heart: [Build #317651](https://buildkite.com/elastic/kibana-pull-request/builds/317651) failed aa0ae050410081d931d40425c53732532391a597
* :broken_heart: [Build #317606](https://buildkite.com/elastic/kibana-pull-request/builds/317606) failed e01f8e50c80d7fd0c9e992f12fc4a84131cc9f33

cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318415","number":318415,"commit":"c0481649aa52ba73a9a7c04e92d60e70ad059cd2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318260","number":318260,"commit":"7cc07a313fbc8284c5daf2cd334edd582dd96903"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317817","number":317817,"commit":"d68908a3677c79a65705b05f6d6af3d2450b2e88"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317665","number":317665,"commit":"4eafaff9051a51776fdac3ed35d8c14a2761a87d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317651","number":317651,"commit":"aa0ae050410081d931d40425c53732532391a597"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317606","number":317606,"commit":"e01f8e50c80d7fd0c9e992f12fc4a84131cc9f33"}],"number":318415}
buildkite-pr-comment-->

mattkime reviewed (approved):
works well and code looks good!

sabarasaba reviewed (approved):
latest lgtm

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/structure-tree

ElenaStoeva committed:
Fix tests and add more

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/structure-tree

ElenaStoeva commented:
/ci

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/structure-tree
